design-proposals/generalize-virt-stack/gen-virt-stack.md

```diff
@@ -0,0 +1,130 @@
+
+Although KubeVirt currently relies on libvirt to create and manage virtual machine instances (VMIs), it relies specifically on the QEMU/KVM virtualization stack (VMM and hypervisor) to host the VMI. This limits KubeVirt from being used in settings where a different VMM or hypervisor is used. 
+
+In fact, libvirt itself is flexible enough to support a diverse set of VMMs and hypervisors. The libvirt API delegates its implementation to one or more internal drivers, dependending on the [connection URI](https://libvirt.org/uri.html) passed when initializing the library. The list of currently supported hypervisor drivers in Libvirt are:
+- [LXC - Linux Containers](https://libvirt.org/drvlxc.html)
+- [OpenVZ](https://libvirt.org/drvopenvz.html)
+- [QEMU/KVM/HVF](https://libvirt.org/drvqemu.html)
+- [VirtualBox](https://libvirt.org/drvvbox.html)
+- [VMware ESX](https://libvirt.org/drvesx.html)
+- [VMware Workstation/Player](https://libvirt.org/drvvmware.html)
+- [Xen](https://libvirt.org/drvxen.html)
+- [Microsoft Hyper-V](https://libvirt.org/drvhyperv.html)
+- [Virtuozzo](https://libvirt.org/drvvirtuozzo.html)
+- [Bhyve - The BSD Hypervisor](https://libvirt.org/drvbhyve.html)
+- [Cloud Hypervisor](https://libvirt.org/drvch.html)
+
+There are several parts in the KubeVirt source-code that hard-code the use of the QEMU/KVM hypervisor driver, which prevents the creation of VMIs using another hypervisor driver. Therefore, KubeVirt needs to be updated to introduce the choice of the backend libvirt hypervisor driver to use for creating a given VMI. This would expand the set of scenarios in which KubeVirt can be used.
+
+## Goals
+
+KubeVirt should be able to offer a choice to its users over which libvirt hypervisor-driver they want to use to create their VMI.
+
+## Non Goals
+
+Support all features available in KubeVirt in all libvirt hypervisor drivers. As libvirt makes progress in bringing feature parity among its hypervisor drivers, KubeVirt will also enable more features in the evolving hypervisor drivers.
+
+## Definition of Users
+
+This proposal is aimed at serving users who intend to use KubeVirt on a cluster where at least one node has a virtualization stack different from the default QEMU-KVM stack.
+
+## User Stories
+
+- A user trying to use KubeVirt on a cluster of machines that have a hypervisor-VMM pair that is not necessarily QEMU/KVM.
+- A regular user of libvirt with any of its supported hypervisor drivers who now wants to leverage the orchestration capability provided by KubeVirt. But the user does not want to abandon their hypervisor driver of choice.
+- A regular user of a hypervisor-VMM-specific orchestration capability to expand their use to a cluster with a diverse set of hypervisor-VMM pairs.
+
+## Repos
+
+- KubeVirt
+- Libvirt
+
+# Design
+
+## API Changes
+
+Addition of a `vmi.hypervisor` field. Example of how a user could request a specific hypervisor as the underlying libvirt hypervisor-driver through the VMI spec:
```

Comment:
VMI seems to be quite fine granular.

if,t hen shouldnÄt it be a cluster level setting?

Reply:
I am considering a scenario in which different cluster nodes could have a different virtualization-stack. In KubeVirt `virt-handlers` on different cluster nodes are independent, so IMO there is no reason to not set hypervisor at this fine granularity.

Reply:
What is the reason for having different hypervisors in a single cluster?

There is also a cluster level impact, i.e. the overhead calculation.

Reply:
No specific reason. Based on my understanding of the KubeVirt code, the overhead calculation is for the virt-launcher pod alone, so one could (in theory) have diff virt-launcher pods with diff hypervisors running on the same cluster. However, I could be wrong, so please correct me.

I don't have a specific scenario in mind as of now that would require multiple hypervisors on the same cluster, but it is a more flexible design choice imo to have the hypervisor-specific logic limited to the components that are tied to specific nodes (i.e., virt-handler and virt-launcher).

Reply:
IMO, it makes sense to allow multiple hypervisors in the same cluster. Different hypervisor could fit to different use cases and we could have a unified management.